### PR TITLE
Dev/ade9153a example

### PR DESCRIPTION
**Pull request by raezt**
## Pull Request Description

Please replace this with a detailed description and motivation of the changes. 
You can tick the checkboxes below with an 'x' between square brackets or just check them after publishing the PR. 
If this PR contains a breaking change, list dependent PRs and try to push all related PRs at the same time.

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [x] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [x] I have build all projects affected by the changes in this PR
- [x] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [ ] I have updated the documentation (wiki pages, ReadMe etc), if applies

### CODE REVIEW

File: drivers/meter/ade9153a/ade9153a.c [hunk 1]
```diff
@@ -0,0 +1,4422 @@
...
+	}
+
+	if (dev->gpio_sck) {
+		ret = no_os_gpio_direction_output(dev->gpio_sck,
+						  NO_OS_GPIO_HIGH);
+		if (ret)
+			return ret;
+	}
+
+	ade9153a_hw_reset(dev);
```
Comment (amiclaus) [inline drivers/meter/ade9153a/ade9153a.c:94]: no return handling.

File: drivers/meter/ade9153a/ade9153a.c [hunk 2]
```diff
@@ -0,0 +1,4422 @@
...
+	}
+
+	if (dev->gpio_sck) {
+		ret = no_os_gpio_direction_output(dev->gpio_sck,
+						  NO_OS_GPIO_LOW);
+		if (ret)
+			return ret;
+	}
+
+	ade9153a_hw_reset(dev);
```
Comment (amiclaus) [inline drivers/meter/ade9153a/ade9153a.c:127]: ditto.

File: drivers/meter/ade9153a/ade9153a.c [hunk 3]
```diff
@@ -0,0 +1,4422 @@
...
+					      init_param.gpio_ss);
+		if (ret)
+			goto error_gpio_reset;
+
+		ret = no_os_gpio_get_optional(&dev->gpio_sck,
+					      init_param.gpio_sck);
+		if (ret)
+			goto error_spi;
+
+		dev->reset_delay = init_param.reset_delay;
```
Comment (amiclaus) [inline drivers/meter/ade9153a/ade9153a.c:251]: can the reset delay have multiple values?

File: drivers/meter/ade9153a/ade9153a.c [hunk 4]
```diff
@@ -0,0 +1,4422 @@
...
+	ret = no_os_gpio_get_optional(&dev->gpio_reset,
+				      init_param.gpio_reset);
+	if (ret)
+		goto error_irq;
+
+	if (dev->gpio_reset) {
+		ret = no_os_gpio_direction_output(dev->gpio_reset, NO_OS_GPIO_HIGH);
+		if (ret)
+			goto error_gpio_reset;
+		//return ret;
```
Comment (amiclaus) [inline drivers/meter/ade9153a/ade9153a.c:236]: drop this.

File: drivers/meter/ade9153a/ade9153a.c [hunk 5]
```diff
@@ -0,0 +1,4422 @@
...
+	if (!dev)
+		return -ENODEV;
+	if (!reg_data)
+		return -EINVAL;
+
+	/* reset command buffer */
+	for (i = 0; i < 8; i++)
+		buff[i] = 0;
+
+	addr = (uint16_t) ((reg_addr << 4) & 0x0FFFF);
```
Comment (amiclaus) [inline drivers/meter/ade9153a/ade9153a.c:407]: use NO_OS_GENMASK + field_prep/get

File: drivers/meter/ade9153a/ade9153a.c [hunk 6]
```diff
@@ -0,0 +1,4422 @@
...
+	/* data to be written */
+	uint32_t data;
+	/* offset of data to be written in the buffer */
+	uint8_t data_byte_offset = 2;
+	int i;
+
+	if (!dev)
+		return -ENODEV;
+
+	addr = (uint16_t) ((reg_addr << 4) & 0x0FFFF);
```
Comment (amiclaus) [inline drivers/meter/ade9153a/ade9153a.c:472]: ditto.

File: drivers/meter/ade9153a/ade9153a.h [hunk 7]
```diff
@@ -0,0 +1,1471 @@
...
+
+// Start autocalibration AV channel
+int ade9153a_start_autocal_av(struct ade9153a_dev *dev,
+			      enum ade9153a_acalmode_e mode);
+
+// Read autocalibration values
+int ade9153a_read_autocal_vals(struct ade9153a_dev *dev,
+			       struct ade9153a_autocal_vals *data);
+
+#endif // __ADE9153A_H__
```
Comment (amiclaus) [inline drivers/meter/ade9153a/ade9153a.h:1471]: new line missing.

File: drivers/meter/ade9153a/ade9153a.h [hunk 8]
```diff
@@ -0,0 +1,1471 @@
...
+/**
+ * @enum ade9153a_pwr_settle_e
+ * @brief ADE9153A Power settle.
+ */
+enum ade9153a_pwr_settle_e {
+	/* Configure the time for the power and filter-based */
+	/* rms measurements to settle before starting the */
+	/* power, energy, and CF accumulations */
+	/* 64 ms settle */
+	ADE9153a_SETTLE_64_MS = 0,
```
Comment (amiclaus) [inline drivers/meter/ade9153a/ade9153a.h:482]: use only uppercase letters for enums

File: drivers/meter/ade9153a/ade9153a.h [hunk 9]
```diff
@@ -0,0 +1,1471 @@
...
+/******************************************************************************/
+
+/**
+ * @enum ade9153a_bi_gain_e
+ * @brief ADE9153A Bi Gain.
+ */
+enum ade9153a_bi_gain_e {
+	/* PGA gain for Current Channel B */
+	/* Gain 1 */
+	ADE9153a_PGA_CHB_GAIN_1 = 0,
```
Comment (amiclaus) [inline drivers/meter/ade9153a/ade9153a.h:431]: if there are only consecutive values in an enum, starting from 0, then there is no need for explicit initialization.

File: drivers/meter/ade9153a/ade9153a.h [hunk 10]
```diff
@@ -0,0 +1,1471 @@
...
+/******************************************************************************/
+
+/* SPI commands */
+#define ADE9153A_SPI_READ				    NO_OS_BIT(3)
+
+/* Lock write key */
+#define ADE9153A_LOCK_KEY					0x3C64
+
+/* Unlock write key */
+#define ADE9153A_UNLOCK_KEY					0x4AD1
```
Comment (amiclaus) [inline drivers/meter/ade9153a/ade9153a.h:72]: fix indentation for all defines.

File: drivers/meter/ade9153a/ade9153a.c [hunk 11]
```diff
@@ -0,0 +1,4422 @@
...
+	data->avcc_reg_val = (int32_t)temp_val;
+
+	// read avcert value
+	ret = ade9153a_read(dev, ADE9153A_REG_MS_ACAL_AVCERT, &temp_val);
+	if (ret)
+		return ret;
+	data->avcert_reg_val = (int32_t)temp_val;
+
+	return 0;
+}
```
Comment (amiclaus) [inline drivers/meter/ade9153a/ade9153a.c:4422]: new line.

File: projects/eval-ade9153a/Makefile [hunk 12]
```diff
@@ -0,0 +1,5 @@
+include ../../tools/scripts/generic_variables.mk
+
+include src.mk
+
+include ../../tools/scripts/generic.mk
```
Comment (amiclaus) [inline projects/eval-ade9153a/Makefile:5]: new line.

File: projects/eval-ade9153a/src.mk [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+	$(PLATFORM_DRIVERS)/maxim_pwm.c
+
+# Application entry point
+SRCS += $(PROJECT)/src/main.c
+INCS += $(PROJECT)/src/common/common_data.h
+SRCS += $(PROJECT)/src/common/common_data.c
+
+# ADE9153A driver files
+INCS += $(DRIVERS)/meter/ade9153a/ade9153a.h
+SRCS += $(DRIVERS)/meter/ade9153a/ade9153a.c
```
Comment (amiclaus) [inline projects/eval-ade9153a/src.mk:65]: new line

File: projects/eval-ade9153a/src/common/common_data.c [hunk 14]
```diff
@@ -0,0 +1,407 @@
...
+	float constant, power_factor_val, frequency_val, angle_ai_av_val;
+	// status of freq select
+	uint8_t status;
+	// skip temp if timeout
+	uint8_t skip_temp;
+
+
+	ret = ade9153a_temp_val(dev, &temp_val);
+	if (ret)
+		if (ret == -ENODATA) {
```
Comment (amiclaus) [inline projects/eval-ade9153a/src/common/common_data.c:192]: if the `if` statements has brackets then the `else` should have too (even if it is only one line)

File: projects/eval-ade9153a/src/common/common_data.c [hunk 15]
```diff
@@ -0,0 +1,407 @@
...
+	no_os_mdelay(50000);
+
+	ret = ade9153a_write(dev, ADE9153A_REG_MS_ACAL_CFG, 0x00000000);
+	if (ret)
+		return ret;
+
+	no_os_mdelay(100);
+
+	return 0;
+}
```
Comment (amiclaus) [inline projects/eval-ade9153a/src/common/common_data.c:407]: new line.

File: projects/eval-ade9153a/src/common/common_data.c [hunk 16]
```diff
@@ -0,0 +1,407 @@
...
+
+	ret = ade9153a_start_autocal_ai(dev, NORMAL);
+	if (ret) {
+		pr_info("ERROR %d \n", ret);
+		return ret;
+	}
+
+	pr_info("IA autocalibration started \n");
+
+	no_os_mdelay(20000);
```
Comment (amiclaus) [inline projects/eval-ade9153a/src/common/common_data.c:382]: some explanation for these delays would be nice :)

File: projects/eval-ade9153a/src/common/common_data.h [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+/* Read measurements */
+int read_measurements(struct ade9153a_dev *dev);
+
+/* Autocalibration read values */
+int autocalibration_read_vals(struct ade9153a_dev *dev);
+
+/* Autocalibration start */
+int autocalibration_start(struct ade9153a_dev *dev);
+
+#endif
```
Comment (amiclaus) [inline projects/eval-ade9153a/src/common/common_data.h:162]: ditto: new line.

File: projects/eval-ade9153a/src/main.c [hunk 18]
```diff
@@ -0,0 +1,238 @@
...
+remove_uart:
+	no_os_uart_remove(uart_desc);
+remove_irq:
+	no_os_irq_ctrl_remove(ade9153a_nvic_desc);
+
+error:
+	pr_err("ERROR\n");
+	return ret;
+
+}
```
Comment (amiclaus) [inline projects/eval-ade9153a/src/main.c:238]: new line.

**Review by amiclaus — COMMENTED** [review]
some small comments on my side.

**Comment by raezt** [comment]
All the comments have been addressed.

File: drivers/meter/ade9153a/ade9153a.c [hunk 19]
```diff
@@ -0,0 +1,4418 @@
...
+	struct ade9153a_dev *desc = dev;
+	/* data read */
+	uint8_t reg_val;
+
+	/* Disable interrupt while reading data. */
+	ret = no_os_irq_disable(desc->irq_ctrl,
+				desc->gpio_rdy->number);
+	if (ret)
+		return;
+
```
Comment (buha) [inline drivers/meter/ade9153a/ade9153a.c:147]: isn't there a data read supposed to be happening here ?

File: drivers/meter/ade9153a/ade9153a.h [hunk 20]
```diff
@@ -0,0 +1,1470 @@
...
+
+/* Version product */
+#define ADE9153A_VERSION						0x9153a
+
+/* Reset Delay */
+#define ADE9153A_RESET_DEL						0x1F4
+
+/* Address range of 16 bit registers */
+#define ADE9153A_START_16ADDR					0x473
+#define ADE9153A_END_16ADDR						0x600
```
Comment (buha) [inline drivers/meter/ade9153a/ade9153a.h:89]: there is something wrong with these TAB indentations, if i use the View Raw file functionality on Github they still get displayed badly, so this is definitely something to fix

File: drivers/meter/ade9153a/ade9153a.h [hunk 21]
```diff
@@ -0,0 +1,1470 @@
...
+	/** operation of sensor */
+	uint8_t							ai_swap;
+	/** AI gain init value */
+	enum ade9153a_ai_gain_e			ai_pga_gain;
+	/** High pass filter corner freq init value */
+	enum ade9153a_hpf_crn_e			hpf_crn;
+	/** Energy accumulation freq select */
+	enum ade9153a_selfreq_e			freq;
+	/** Vlevel value */
+	uint32_t						vlevel;
```
Comment (buha) [inline drivers/meter/ade9153a/ade9153a.h:702]: same comment about indentation applies to these structures, i usually avoid such indentations in structures and just use a single whitespace instead

File: drivers/meter/ade9153a/ade9153a.c (hunk 19 above)
Comment (raezt) [reply drivers/meter/ade9153a/ade9153a.c:147]: In the example the data read is not interrupt based. I've left the irq handler in the driver for user applications development. Maybe I should add a comment like this "// user code goes here"

**Comment by raezt** [comment]
v3

- fixed indentation issues.

**Review by amiclaus — COMMENTED** [review]
now I see, the builds.json is missing, therefore the project is not included in the build system.

**Comment by raezt** [comment]
v4

- added builds.json

**Review by amiclaus — APPROVED** [review]

**Pull request merged**